File: solutions/1790-lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        p_copy, q_copy = p, q
        while p_copy != q_copy:
            if q_copy:
                q_copy = q_copy.parent
            else:
                q_copy = p
            if p_copy:
                p_copy = p_copy.parent
            else:
                p_copy = q
        return p_copy
            

File: solutions/1790-lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.java
class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        Set<Node> set = new HashSet<>();
        while(p != null || q != null) {
            if(p != null) {
                if(!set.add(p)) {
                    return p;
                }
                p = p.parent;
            }
            if(q != null) {
                if(!set.add(q)) {
                    return q;
                }
                q = q.parent;
            }
        }
        return null;
    }
}